packages/cli/src/commands/deploy-v2.ts

```diff
@@ -0,0 +1,111 @@
+import glob from "glob";
+import { basename } from "path";
+import type { CommandModule } from "yargs";
+import { loadWorldConfig, ResolvedWorldConfig } from "../config/loadWorldConfig.js";
+import { deploy } from "../utils/deploy-v2.js";
+import { getOutDirectory } from "../utils/forgeConfig.js";
+
+type Options = {
+  configPath?: string;
+  print?: boolean;
+};
+
+const commandModule: CommandModule<Options, Options> = {
+  command: "deploy-v2",
+
+  describe: "Deploy MUD v2 contracts",
+
+  builder(yargs) {
+    return yargs.options({
+      configPath: { type: "string", desc: "Path to the config file" },
+      print: { type: "boolean", desc: "Print the resolved config" },
+    });
+  },
+
+  async handler({ configPath, print }) {
+    const config = await loadWorldConfig(configPath);
+
+    // Run forge build
+    try {
+      await execa("forge", ["build"], { stdout: "inherit", stderr: "pipe" });
+    } catch (error: any) {
+      if (error?.stderr) {
+        console.error(error.stderr);
+      }
+      console.error(chalk.red("Error running `forge build`"));
+      process.exit(1);
+    }
+
+    // Get a list of all contracts in output ending with `System`
+    const outDir = await getOutDirectory();
+    const defaultSystemNames = glob
+      // Get all files in outDir ending in `*System.sol  `
+      .sync(`${outDir}/*System.sol`)
+      // Get the basename of the file
+      .map((path) => basename(path, ".sol"))
+      // Ignore the base System contract
+      .filter((name) => name !== "System");
+
+    // Include overridden systems (even if they don't end with `System`)
+    const allSystemNames = [...new Set([...defaultSystemNames, ...Object.keys(config.overrideSystems)])];
+
+    // Exclude systems that are in the exclude list
+    const systemNames = allSystemNames.filter((name) => !config.excludeSystems.includes(name));
+
+    // Get all contract names
+    const allContractNames = glob
+      // Get all files in outDir ending in `*.sol  `
+      .sync(`${outDir}/*.sol`)
+      // Get the basename of the file
+      .map((path) => basename(path, ".sol"));
+
+    // Verify that the list of system names is valid
+    for (const systemName of systemNames) {
+      if (!allContractNames.includes(systemName)) {
+        console.log(chalk.red(`Can not find system "${systemName}"`));
+        process.exit(1);
+      }
+    }
+
+    // Verify that the access lists only include existing systems
+    for (const [systemName, system] of Object.entries(config.overrideSystems)) {
+      if (system.openAccess) continue;
+      for (const accessElement of system.accessList) {
+        if (!ethers.utils.isAddress(accessElement) && !systemNames.includes(accessElement)) {
+          console.log(
+            chalk.red(
+              `Invalid access list in "${systemName}": "${accessElement}" is not a valid Ethereum address or name of a local system.`
+            )
+          );
+          process.exit(1);
+        }
+      }
+    }
```

Comment:
I wonder if most of the logic in this function should be moved into the `zod` validators. I put it here for now because it depends on running `forge build` and then grabbing all `*System.sol` contracts from the file system, which felt like something that doesn't really fit into `zod`. Thoughts @holic @dk1a?

Reply:
I agree it doesn't fit into the world's schema.
But you could add a separate function that accepts `loadWorldConfig`'s output and the results that depend on `forge`. The main benefit there is not so much a zod schema though, but having that function separated from cli

Reply:
Kinda like how my `tablegen` command does just some filesystem writing, and `renderTablesFromConfig` is a separate exported function you can use on its own

Reply:
`renderTablesFromConfig` also doesn't really do any validation (you would need some of course), but I think you don't need to validate world config again. You can just export the parseConfig function as well

Reply:
Another thought I just had is we could pass `existingContractNames` as array into `loadWorldConfig`, so it can be used for validation inside `zod` (checking that the system overrides and excludes only refer to existing contracts). And we could probably use something like `zod`'s `transform` to resolve `accessList` into `accessListSystems` and `accessListAddresses`. Downside is `forge build` would happen before validating the config, but it should be really quick if no contracts change in between the config changes.

Reply:
I defer to @dk1a, who now has more zod experience than me haha

Reply:
> Another thought I just had is we could pass existingContractNames as array into loadWorldConfig, so it can be used for validation inside zod

I haven't really explored your code yet, but it doesn't seem like `existingContractNames` should be a part of the config loader (unless I'm missing something and it's a config parameter).

To summarize: I'm suggesting another little zod schema and a separate method that accepts config+forge stuff, and validates/transforms those lists. And it can assume world config is otherwise validated already